libgnucash/engine/Account.cpp

```diff
@@ -3397,10 +3403,16 @@ GetBalanceAsOfDate (Account *acc, time64 date, gboolean ignclosing)
             /* Since lp is now pointing to a split which was past the reconcile
              * date, get the running balance of the previous split.
              */
-            if (ignclosing)
-                balance = xaccSplitGetNoclosingBalance( (Split *)lp->prev->data );
-            else
+            switch (baltype) {
+            case 0:
                 balance = xaccSplitGetBalance( (Split *)lp->prev->data );
+                break;
+            case 1:
+                balance = xaccSplitGetNoclosingBalance( (Split *)lp->prev->data );
+                break;
+            case 2:
+                balance = xaccSplitGetReconciledBalance( (Split *)lp->prev->data );
+            }
```

Comment:
This doesn't do what you want because it operates on `date_posted` instead of `reconcile_date`.  You can't assume that a split posted before a statement date was necessarily reconciled in that statement. If it was reconciled in the next statement, as is likely, then xaccAccountRecomputeBalances will treat it as balanced and the resulting reconciled balance won't match the statement's opening balance.

Reply:
Explained in https://bugs.gnucash.org/show_bug.cgi?id=797640#c11 why I don't think the reconcile_date matters here.

Reply:
You explained why you don't understand: You're laboring under the misconception that all transactions posted in a particular month clear the bank that month.

Reply:
Now I know `reconcile_date` means `statement_date` I've amended `xaccAccoutGetReconciledBalanceAsOfDate` to find the _last_ split whose reconciled date fits, and return its reconciled balance. If this last attempt is still not right then I am not sure whether I can pursue this further.